Rewrite and extend computer graphics notes to follow a more conventional style (#3)
VIK-CE-Notes/vik-ce-notes.github.io: docs/notes/sem4/computer_graphics/lectures/2.md

??? example Bővebben a VBO/VAO-ról...
    A VBO és VAO megértését segítő [videó](https://www.youtube.com/watch?v=Rin5Cp-Hhj8).

A `glDrawArrays` függvény juttatja el ezeket a VBO/VAO-kat a GPU-hoz. Ezek floating point regiszterekbe kerülnek. A vertex shader meghatározza a normalizált eszközkoordináta-rendszert ($(-1, -1)$ és $(1, 1)$ között). Ezután a `gl_Position` tárolja a normalizált eszközkoordinátákat, az itt lévő regiszterek pedig például színértéket tárolhatnak.

!!! note ""
    A `gl_Position`-ben tárolt értékek mindig 4 dimenziósak lesznek és a 3D-s projektív geometriai szabályok szerint lesznek értelmezve.

<!-- ez az állítás nem teljes így 1. mert beadhatok neki vec2/3-at és akkor nagyon jól érzem magam, de valójában akkor is vec4 lesz, 2. persze lehet csinálni másmilyen geometriát is, de a GPU ettől még nem fogja úgy számolni nem feltétlenül úgy fogja számolni -->

> Ezt követően fix műveleti egységek mennek végbe (vágás, raszterizáció, stb.). Ezek optimalizált hardware műveletek.

Reviewer: ```suggestion
Ezt követően fix műveleti egységek mennek végbe (vágás, raszterizáció, stb.). Ezek optimalizált hardver műveletek.
```

- Vágás: kidobjuk ami "nem látszik".

> - Viewport transzformáció: A normalizált eszközkoordinátákból a képernyő koordináta-rendszerévé ($(0, 0)$ és pl. $(600, 800)$ között) való transzformálás.

Reviewer: ```suggestion
- Viewport transzformáció: A normalizált eszközkoordinátákból a képernyő koordináta-rendszerévé ($(0, 0)$ és pl. $(800, 600)$ között) való transzformálás.
```
de ez már nagyon kukacoskodás

- Raszterizáció: Egy adott sokszög csúcspontjai ismeretében meghatározza az összes olyan pixelt ami benne van a sokszögben, interpolálja a regiszter értékeket (pl. szín).

> Ezeket követően már konkrét pixelekkel dolgozik a GPU. A fragment shader a pixelek tényleges színét határozza meg. Végezetül kompozitálás után a pixelek bekerülnek a rasztertárba, ahonnan már egyenesen a képernyőre kerülnek. A kompozitálás azért fontos, hiszen itt dől el, hogy melyik pixelek lennének más pixelek "mögött", az átlátszóság kezeléséhez kiemelten fontos.

Reviewer: ```suggestion
Ezeket követően már konkrét pixelekkel dolgozik a GPU. A fragment shader a pixelek tényleges színét határozza meg. Végezetül kompozitálás után a pixelek bekerülnek a rasztertárba (frame buffer), ahonnan már egyenesen a képernyőre kerülnek. A kompozitálás azért fontos, mert itt dől el, hogy melyik pixelek lennének más pixelek "mögött", ami az áttetszőség és a takarás kezeléséhez elengedhetetlen.
```

> ahonnan már egyenesen a képernyőre kerülnek

bár ilyen egyszerű lenne xd... de nem baj ne vigyük le szegényeket a compositor rabbitholeon

### Primitívek

Tételezzük fel, hogy tetszőleges pontunk a VAO-ban (legyenek ezek $x_i$-vel jelölve, ahol $i \in \N $). A `gl_DrawArrays` hogyan fogja őket értelmezni? Tekintsük először a függvény paramétereit:

```cpp
glDrawArrays(primitiveType, startIdx, numOfElements);
```

A `startIdx` és `numOfElements` magáért beszélnek, viszont a `primitiveType`-ról részletesebben szót kell ejteni. Ez a jegyzet 7 darab ilyen primitívet tárgyal, ezek pedig:

- `GL_POINTS`: külön pontokként jeleníti meg a VAO pontjait
- `GL_LINES`: kettessével összeköti a megadott pontokat, így például az alábbi _szakaszok_ jönnének létre: $(x_0, x_1), (x_2, x_3), ...$
- `GL_LINE_STRIP`: az adott pontot mindig a következő ponttal köti össze: $(x_0, x_1), (x_1, x_2), (x_2, x_3), ...$

> - `GL_LINE_LOOP`: ugyan az mint a `GL_LINE_STRIP`, csak a legutolsó pontot a legelsővel köti össze

Reviewer: ```suggestion
- `GL_LINE_LOOP`: ugyan az mint a `GL_LINE_STRIP`, csak a legutolsó pontot a legelsővel is összeköti
```

<!-- 

NOTE: Az alábbi jegyzetet az előadáson készítettem, de nem tartottam annyira fontosnak, hogy a végső jegyzetbe bele írjam. 

Mindenesetre kommentként meghagyom, ha valaki gondolja akkor kiegészítheti és bekerülhet a jegyzetbe. - levy04

---

Zöld háromszög program:
feladatnál a `glApp` osztályból származunk le a saját "x_app"-unkat.
shadereket GLSL nyelven programozzuk (C jellegű)

sokszor kell ugyan azt: erre van a framework.h és framework.cpp: ezekbe kiszervezzük a közös dolgokat

példa: viktor vasarely (Vásárhelyi Győző) festményei: sok optikai trükk van bennük.
- betüremkedő gömb: körök mozognak fel le, teljesen 2D. 1 körünk van, azt transzformálgatjuk. a középpont középen kezdődik, először felfele megy, aztán vissza középre (sin hullámzás). eközben nő a kör sugara, változik a színe. az illúzióhoz a középpont is mozog fel le.

- konvex burok, interakció (legkisebb olyan sokszög, ami az összes pontot tartalmazza). a legalsó pontból indulunk, felveszünk egy vízszintes egyenes. a kövi pont az lesz, amihez a legkevesebbet kell elforgatni ezt az egyenest.

- tessaract (4D kocka), animáció.

-->

## Input kezelése

> Az operációs rendszernél a koordináta-rendszer origója a bal felső sarokban található, az $y$ tengely fejjel lefelé van! A `glViewport` viszont a bal alsó sarkot tekinti origónak, tehát valahogy így néz ki a dolog:

Reviewer: ```suggestion
Az operációs rendszer koordináta-rendszerének origója a bal felső sarokban van, tehát az $y$ tengely **fejjel lefelé** van, "$y$ lefele mutat". A `glViewport` viszont a **bal alsó** sarkot tekinti origónak, és az $y$ tengely "felfele mutat". Valahogy így kell elképzelni:
```

    - [ ] A GPU pixel árnyaló programja eldönti, hogy melyik pixelt színezze ki a kért színre.
        - Magyarázat: Az OpenGL egy állapotgép, nem "dönt" semmiről.

---

!!! question 7\. Válassza ki a helyes állításokat az OpenGL körrajzoló képességével kapcsolatban.
    - Az OpenGL nem tud kört rajzolni, mert a művelet felesleges, hiszen a kör közelíthető szabályos sokszöggel.
    - Az OpenGL nem tud kört rajzolni, mert a GPU fejlesztők nem ismerték a kör egyenletét.
    - Az OpenGL nem tud kört rajzolni, mert projektív geometriában nincs távolság, ezért nincs kör sem, helyette keletek lehetnek, azok viszont túl bonyolultak lennének a vágás és raszterizáció hw. implementációjához.
    - Az OpenGL nem tud kört rajzolni, hogy kitoljon a programozókkal.
    - Az OpenGL-nek a kör középpontját és sugarát kell átadni, hogy ki tudja rajzolni.

??? tip Megoldás
    - [x] Az OpenGL nem tud kört rajzolni, mert a művelet felesleges, hiszen a kör közelíthető szabályos sokszöggel.
        - Magyarázat: Lásd [fentebb](#funkcionális-modell)
    - [ ] Az OpenGL nem tud kört rajzolni, mert a GPU fejlesztők nem ismerték a kör egyenletét.

>         - Magyarázat: Szerintem a $8$ általánost elvégeszték azért a GPU fejlesztők.

Reviewer: xd
